panther_battery/src/adc_node.cpp

```diff
@@ -0,0 +1,213 @@
+    std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 0, 0),
+    std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 2, 0), default_battery_params);
+
+  battery_count_ = battery_2_->Present() ? 2 : 1;
+  if (battery_count_ == 2) {
+    RCLCPP_INFO(this->get_logger(), "Second battery detected");
+    battery_1_ = std::make_unique<Battery>(
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc1_reader_, 0, 0),
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc1_reader_, 2, adc_current_offset_),
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 1, 0),
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 3, 0), default_battery_params);
+    battery_2_pub_ = this->create_publisher<BatteryStateMsg>("battery_2_raw", 10);
+  } else {
+    battery_1_ = std::make_unique<Battery>(
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc1_reader_, 0, 0),
+      [&]() {
+        return adc1_reader_->GetADCMeasurement(2, adc_current_offset_) +
+               adc1_reader_->GetADCMeasurement(1, adc_current_offset_);
+      },
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 1, 0),
+      std::bind(&ADCDataReader::GetADCMeasurement, *adc0_reader_, 3, 0), default_battery_params);
+    battery_2_.reset();
+  }
+
+  battery_pub_ = this->create_publisher<BatteryStateMsg>("battery", 10);
+  battery_1_pub_ = this->create_publisher<BatteryStateMsg>("battery_1_raw", 10);
+
+  io_state_sub_ = this->create_subscription<IOStateMsg>(
+    "hardware/io_state", 10,
+    [&](const IOStateMsg::SharedPtr msg) { charger_connected_ = msg->charger_connected; });
+
+  // running at 10 Hz
+  battery_pub_timer_ = this->create_wall_timer(
+    std::chrono::milliseconds(100), std::bind(&ADCNode::BatteryPubTimerCB, this));
+
+  RCLCPP_INFO(this->get_logger(), "Node started");
+}
+
+void ADCNode::BatteryPubTimerCB()
+{
+  auto header_stamp = this->get_clock()->now();
+
+  try {
+    battery_1_->Update(header_stamp, charger_connected_);
+    if (battery_count_ == 2) {
+      battery_2_->Update(header_stamp, charger_connected_);
+    }
+    last_battery_info_time_ = this->get_clock()->now();
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR(this->get_logger(), "Error reading battery data: %s. ", e.what());
+  }
+
+  if (
+    (this->get_clock()->now() - last_battery_info_time_) >
+    rclcpp::Duration::from_seconds(battery_timeout_)) {
+    battery_1_->Reset(header_stamp);
+    if (battery_count_ == 2) {
+      battery_2_->Reset(header_stamp);
+    }
+  }
+
+  if (battery_count_ == 2) {
+    battery_pub_->publish(
+      MergeBatteryMsgs(battery_1_->GetBatteryMsg(), battery_2_->GetBatteryMsg()));
+    battery_1_pub_->publish(battery_1_->GetBatteryMsgRaw());
+    battery_2_pub_->publish(battery_2_->GetBatteryMsgRaw());
+  } else {
+    battery_pub_->publish(battery_1_->GetBatteryMsg());
+    battery_1_pub_->publish(battery_1_->GetBatteryMsgRaw());
+  }
```

Comment:
Battery message object could be created only once and later passed to the publishers

Reply:
What's the point in creating the battery msg object that is used only once?